pkg/arvo/lib/lm.hoon

```diff
@@ -0,0 +1,702 @@
+  ++  mmul
+    ~/  %mmul
+    |=  [u=@lms v=@lms]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  mn  `(list @ud)`(shape u)
+    =/  mu  `@ud`-:mn
+    =/  nu  `@ud`+<:mn
+    =/  np  `(list @ud)`(shape v)
+    =/  nv  `@ud`-:np
+    =/  pv  `@ud`+<:np
+    ?.  =(nu nv)  !!  :: make sure this is a valid operation
+    =/  w  `@lms`(zeros mu pv)
+    =/  i  1
+    =/  j  1
+    |-  ^-  @lms
+      ?:  (gth i mu)  `@lms`w
+      ?:  (gth j pv)
+        $(i +(i), j 1, w w)
+      $(i i, j +(j), w `@lms`(set w i j (inner:lvs (getr u i) (getc v j))))
+  ::
+  ::    Matrix exponentiation (A**N, not e(A))
+  ++  mpow
+    ~/  %mpow
+    |=  [u=@lms i=@ud]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  mnu  `(list @ud)`(shape u)
+    =/  mu  `@ud`-:mnu
+    =/  nu  `@ud`+<:mnu
+    =/  w  `@lms`u
+    =/  count  1
+    |-  ^-  @lms
+      ?:  (gth count mu)  `@lms`w
+    $(count +(count), w (mmul u w))  :: TODO check this in Hoon
+  ++  norm  !!
+  ::
+  ::    Matrix trace (sum of main diagonal elements); square matrices only
+  ++  trace
+    ~/  %trace
+    |=  [u=@lms]  ^-  @rs
+    ~_  leaf+"lagoon-fail"
+    =/  mnu  `(list @ud)`(shape u)
+    =/  mu  `@ud`-:mnu
+    =/  nu  `@ud`+<:mnu
+    ?.  =(mu nu)  !!  :: make sure this is a valid operation (square matrix)
+    =/  sum  .0
+    =/  count  1
+    |-  ^-  @rs
+      ?:  (gth count mu)  sum
+    $(count +(count), sum (add:rs sum (get u count count)))
+  ::
+  ::    Operations related to matrix inversion
+  ::    As with matrix multiplication, we're opting for clarity, not efficiency.
+  ++  submatrix
+    ~/  %submatrix
+    |=  [u=@lms ia=@ud ib=@ud ja=@ud jb=@ud]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  mnu  `(list @ud)`(shape u)
+    =/  mu  `@ud`-:mnu
+    =/  nu  `@ud`+<:mnu
+    ?:  (lth ia 1)  !!
+    ?:  (lth ja 1)  !!
+    ?:  (gth ia mu)  !!
+    ?:  (gth ia nu)  !!
+    ?:  (gth ia ib)  !!
+    ?:  (gth ja jb)  !!
+    =/  mw  +((sub ib ia))
+    =/  nw  +((sub jb ja))
+    =/  w  `@lms`(zeros mw nw)
+    =/  ii  ia
+    =/  jj  ja
+    =/  mi  1
+    =/  mj  1
+    |-  ^-  @lms
+      ?:  =(ii +(ib))  w
+      ?:  =(jj +(jb))  $(ii +(ii), jj ja, mi +(mi), mj 1)
+    $(ii ii, jj +(jj), mi mi, mj +(mj), w (set w mi mj (get u ii jj)))
+  ++  augment
+    ~/  %augment
+    |=  [u=@lms]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  mnu  `(list @ud)`(shape u)
+    =/  mu  `@ud`-:mnu
+    =/  nu  `@ud`+<:mnu
+    ?.  =(mu nu)  !!  :: make sure this is a valid operation (square matrix)
+    =/  w  `@lms`(zeros mu (mul mu 2))
+    =/  count  1
+    |-  ^-  @lms
+      ?:  (gth count mu)  `@lms`w
+      =/  ir  (snap (reap mu .0) (dec count) .1)
+      =/  wl  (make:lvs (weld (unmake:lvs (getr u count)) ir))
+    $(count +(count), w (setr w count wl))
+  ::
+  ::    Inverse of positive definite symmetric matrix, per Bauer & Reinsch 1971.
+  ++  invert
+    ~/  %invert
+    |=  [u=@lms]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    (gauss-elim u)
```

Comment:
```suggestion  
    =+  [m n]=[&1 &2]:(shape u)
    (submatrix [1 m] [+(n) (mul 2 n)] (gauss-elim u))
```
per the jet. Otherwise you get the whole augmented matrix
```
> =,(lms.lm =/(a (make ~[.0.5 .0] ~[.0 .0.5] ~) (unmake (invert a))))
~[
  ~[.1 .0 .2 .0]
  ~[.0 .1 .0 .2]
]
```

Reply:
:)